widgetssdk/src/main/java/com/glia/widgets/view/head/controller/ApplicationChatHeadLayoutController.kt

```diff
@@ -0,0 +1,212 @@
+        onEngagementEndUseCase.execute { engagementEnded() }
+        onCallVisualizerEndUseCase.execute(object : GliaOnCallVisualizerEndUseCase.Listener {
+            override fun engagementEnded() {
+                engagementEnded()
+            }
+        })
+        addVisitorMediaStateListenerUseCase.execute(this)
+    }
+
+    override fun onResume(viewName: String) {
+        setResumedViewName(viewName)
+    }
+
+    fun onPause(viewName: String) {
+        clearResumedViewName(viewName)
+    }
+
+    private fun setResumedViewName(viewName: String) {
+        if (resumedViewName == null) {
+            resumedViewName = viewName
+        }
+    }
+
+    private fun clearResumedViewName(viewName: String) {
+        if (isResumedView(viewName)) {
+            resumedViewName = null
+        }
+    }
+
+    private fun isResumedView(viewName: String): Boolean {
+        return resumedViewName != null && resumedViewName == viewName
+    }
+
+    private fun newEngagementLoaded() {
+        state = State.ENGAGEMENT
+        engagementDisposables.add(
+            getOperatorFlowableUseCase.execute()
+                .subscribe(
+                    { operator: Operator -> operatorDataLoaded(operator) }
+                ) { throwable: Throwable ->
+                    Logger.e(
+                        TAG,
+                        "getOperatorFlowableUseCase error: " + throwable.message
+                    )
+                }
+        )
+        updateChatHeadView()
+    }
+
+    private fun engagementEnded() {
+        state = State.ENDED
+        operatorProfileImgUrl = null
+        unreadMessagesCount = 0
+        engagementDisposables.dispose()
+        updateChatHeadView()
+    }
+
+    private fun onUnreadMessageCountChange(count: Int) {
+        unreadMessagesCount = count
+        updateChatHeadView()
+    }
+
+    private fun operatorDataLoaded(operator: Operator) {
+        operatorProfileImgUrl = Utils.getOperatorImageUrl(operator)
+        updateChatHeadView()
+    }
+
+    private fun updateChatHeadViewState(view: ChatHeadLayoutContract.View?) {
+        when (state) {
+            State.ENGAGEMENT -> decideOnEngagementBubbleDesign(view)
+            State.QUEUEING -> view?.showQueueing()
+            State.ENDED -> view?.showPlaceholder()
+        }
+    }
+
+    private fun decideOnEngagementBubbleDesign(view: ChatHeadLayoutContract.View?) {
+        if (isCallVisualizerUseCase()) {
+            view?.showScreenSharing()
+        } else if (operatorProfileImgUrl != null) {
+            view?.showOperatorImage(operatorProfileImgUrl)
+        } else {
+            view?.showPlaceholder()
+        }
+    }
+
+    private fun updateOnHold() {
+        if (isOnHold && state == State.ENGAGEMENT) {
+            chatHeadLayout?.showOnHold()
+        } else {
+            chatHeadLayout?.hideOnHold()
+        }
+    }
+
+    private enum class State {
+        ENDED, QUEUEING, ENGAGEMENT
+    }
+
+    companion object {
```

Comment:
I would suggest having companion object at the very top of the class, then any static variables and functions are always in one place

Reply:
According to [Kotlin coding conventions](https://kotlinlang.org/docs/coding-conventions.html#class-layout) it is placed in the right place.

Reply:
True, that is why I wrote that I suggest the different order. This in my mind is the single most wrong thing about Kotlin. The reasoning is that the variables are declared on the top, why would static variables and static functions be declared at the very bottom?

Reply:
> Put nested classes next to the code that uses those classes. If the classes are intended to be used externally and aren't referenced inside the class, put them in the end, after the companion object.

It's a bit different than `static classes and functions` and I think the reason is hidden in the above rule, so the rule is not wrong.
I think if the constant variables or functions are intended to be used internally, then they shouldn't be inside the `companion object`, and if they are intended to be used externally then the bottom of the class is the best place for them.